Input - add get_action_peak_strength()

lawnjelly opened the pull request:
This PR records the peak strength over each press, which can be read after release, and therefore enables the user to know how hard button was pushed in short presses. This was previously impossible to read after short presses, because strength is reset to zero on release.

Following from #77055 , I'm including this in a separate PR as it deserves separate scrutiny as it is a subtle issue.

## Discussion
Although #77055 allows detection of quick action presses and releases (for jump keys etc), there is one situation which it doesn't solve, that is, getting the action `strength` and `raw_strength` of these quick presses. The problem is that the existing code sets `action_strength` and `raw_action_strength` to ZERO on release. While this works in most situations, with a quick release, it means that if you attempt to find the `strength` of a button press, the result will be zero, and this information has effectively been lost.

This PR solves the problem by introducing a new function to get the peak strength of an action, which is not reset on release, and can thus be read after a short press (after release).

Storing the peak strength should provide a more reliable measure of how hard the button was pressed, in case several strengths are updated to an action before release.

## Notes
* Again this will require some bikeshedding, there are a number of ways of solving this problem.

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

lawnjelly commented:
Think of it like pressing a jump button, and it registers how hard you've pressed it. Or a fighting game that registers how hard you press punch button or something.

I did initially add this as two extra functions as you describe, and that would be fine too. Maybe there's an even better way of passing this info (like combining it in a `is_action_just_pressed()` type function maybe?). :thinking:

groud commented:
> Think of it like pressing a jump button, and it registers how hard you've pressed it. Or a fighting game that registers how hard you press punch button or something.

I don't deny that it can be useful to know how hard a button has been pressed. And I think some controllers already provide that. My only concern is about it is in the context of a less-than-one-frame event. I think that, in such a case, it might not be very relevant. 

Like, as a dev, if you are interested in the transition from 'not pressed' to 'pressed', I am not sure how relevant would be a strength for such a small amount of time, as users would not really be able to control the pressure anyway. For more than one frame, you can, for example, smooth the strength over several frames to let users control the pressure, but for a single frame, the value might be quite irrelevant IMO.

> I did initially add this as two extra functions as you describe, and that would be fine too. Maybe there's an even better way of passing this info (like combining it in a is_action_just_pressed() type function maybe?). thinking

That's an issue with several things in the API. There's no real way to return structs by value. We may be able to return a RefCounted object but that's not really efficient I guess ? (I say that because it's not used often in the API, so TIWAGOS). In the end I guess two new functions would be better and probably more consistent that most of our APIs.

lawnjelly commented:
> I don't deny that it can be useful to know how hard a button has been pressed. And I think some controllers already provide that. My only concern is about it is in the context of a less-than-one-frame event. I think that, in such a case, it might not be very relevant.

We must consider that not input will be at 60fps. At a low frame rate, or during a slow frame, or at low physics tick rate (e.g. I use 10/15 tps for some games with interpolation) where `is_action_just_pressed()` is used within physics tick, it is very easy for a short press to be within the frame or tick.

I'm actually coming round to the idea that storing a peak strength per action might be more useful for this, as if a button only supports pressure, it removes the problem of a spurious low pressure input following the max pressure of a press, so I'll change the PR to do this, and add a separate function instead of the extra parameter. Getting a peak strength for an action could also be potentially useful outside of the short press scenario.

> That's an issue with several things in the API. There's no real way to return structs by value.

Yup it can be a pain. For a button press actually I don't think there would be a massive need for `raw_strength` as there is no dead zone.

Maybe a new function `get_action_peak_strength()` could be the answer? :thinking: 

I'll try this, but again, depends on whether there is API support. I'll try this out on my gamepad.

lawnjelly committed:
Input - add get_action_peak_strength()

This PR records the peak strength over each press, which can be read after release, and therefore enables the user to know how hard button was pushed in short presses. This was previously impossible to read after short presses, because strength is reset to zero on release.
